<div dir=rtl style="direction:rtl;line-height:300%;">
	<font face="XB Zar" size=5>
		<div align=center>
			<font face="vazir" size=5>
                «قشر کتابخوان»
			</font>
		</div>
    </font>
</div>

<br />
<div dir=rtl id="start" style="direction:rtl;line-height:200%;">
	<font face="vazir" size=4>
        صابر بعد از این که دید همه‌ی دوستاش کتاب می‌خونن و در مورد اونا صحبت می‌کنن، تصمیم گرفت خودش هم چند تا کتاب بخونه تا حرفی برای گفتن داشته باشه.
        <br>
        از اون جایی که صابر خودش نمی‌دانست کدام کتاب‌ها بهتر هستند تصمیم گرفت تمام اطلاعات سایت goodreads را دانلود کرده و از شما بخواهد آن‌ها را تحلیل کنید.        
    </font>
</div>

In [11]:
import pandas as pd
import numpy as np


<br />
<div dir=rtl id="start" style="direction:rtl;line-height:200%;">
	<font face="vazir" size=4>
        برای کمک به صابر ابتدا ۴ دیتاست موجود در فایل زیپ را لود کنید و در متغیر‌های زیر بریزید
    </font>
</div>

In [12]:
df = pd.read_csv("./books_db.csv")
authors = pd.read_csv("./authors.csv")
language =pd.read_csv("./language.csv")
publisher = pd.read_csv("./publisher.csv")





<br />
<div dir=rtl id="start" style="direction:rtl;line-height:200%;">
	<font face="vazir" size=4>
        همان‌طور که می‌بینید در دیتاست اصلی (کتاب)، سه متغیر author, lang و publisher تنها به صورت id هستند
        <br>
        برای استفاده بهتر از کل اطلاعات، سه دیتابیس دیگر را با توجه به شماره نمایه‌ی آن‌ها به این دیتابیس اضافه کنید
    </font>
</div>

In [13]:
#todo: about 3 lines
df=pd.merge(df,authors,left_on="author_id",right_index=True )
df=pd.merge(df,language,left_on="lang_id",right_index=True )
df=pd.merge(df,publisher,left_on="publisher_id",right_index=True )
df.head()

bookID                                              title  \
0          1  Harry Potter and the Half-Blood Prince (Harry ...   
1          2  Harry Potter and the Order of the Phoenix (Har...   
3          5  Harry Potter and the Prisoner of Azkaban (Harr...   
2837   10546               The Long-Lost Map (Ulysses Moore #2)   
6190   23300               The Door to Time (Ulysses Moore  #1)   

      average_rating        isbn        isbn13  num_pages  ratings_count  \
0               4.57   439785960  9.780440e+12        652        2095690   
1               4.49   439358078  9.780440e+12        870        2153167   
3               4.56  043965548X  9.780440e+12        435        2339585   
2837            4.00  043977439X  9.780440e+12        272           1693   
6190            3.93   439774381  9.780440e+12        239           3099   

      text_reviews_count publication_date  author_id  lang_id  publisher_id  \
0                  27591        9/16/2006          0        0             0   
1                  29221         9/1/2004          0        0             0   
3                  36325         5/1/2004          0        0             0   
2837                  76         7/1/2006       1280        0             0   
6190                 241         1/1/2006       1280        0             0   

                       name_x code           name_y  
0                J.K. Rowling  eng  Scholastic Inc.  
1                J.K. Rowling  eng  Scholastic Inc.  
3                J.K. Rowling  eng  Scholastic Inc.  
2837  Pierdomenico Baccalario  eng  Scholastic Inc.  
6190  Pierdomenico Baccalario  eng  Scholastic Inc.


<br />
<div dir=rtl id="start" style="direction:rtl;line-height:200%;">
	<font face="vazir" size=4>
        حال که تمام مقادیر اضافه شده‌اند، نیازی به سه ستون id که به عنوان کلید استفاده کردیم نداریم پس آن‌ها را با استفاده از دستور drop حذف می‌کنیم.
        <br>
        * در فصل بعد نحوه‌ی کار با این دستور را یاد می‌گیرید.
    </font>
</div>

In [14]:
df.drop(columns = ['author_id', 'lang_id', 'publisher_id'], inplace = True)


<br />
<div dir=rtl id="start" style="direction:rtl;line-height:200%;">
	<font face="vazir" size=4>
        نام ستون‌ها را طوری تغییر دهید که جدول نهایی چیزی شبیه به جدول زیر باشد:
    </font>
</div>



<table border="1" class="dataframe">
       <thead>
         <tr style="text-align: right;">
           <th></th>
           <th>bookID</th>
           <th>title</th>
           <th>average_rating</th>
           <th>isbn</th>
           <th>isbn13</th>
           <th>num_pages</th>
           <th>ratings_count</th>
           <th>text_reviews_count</th>
           <th>publication_date</th>
           <th>author</th>
           <th>lang</th>
           <th>publisher</th>
         </tr>
       </thead>
       <tbody>
         <tr>
           <th>0</th>
           <td>1</td>
           <td>Harry Potter and the Half-Blood Prince (Harry ...</td>
           <td>4.57</td>
           <td>439785960</td>
           <td>9.780440e+12</td>
           <td>652</td>
           <td>2095690</td>
           <td>27591</td>
           <td>9/16/2006</td>
           <td>J.K. Rowling</td>
           <td>eng</td>
           <td>Scholastic Inc.</td>
           </tr>
    </tbody>
    </table>
           

In [15]:
df.columns=["bookID","title","average_rating","isbn","isbn13","num_pages","ratings_count","text_reviews_count","publication_date","author","lang","publisher"]



<br />
<div dir=rtl id="start" style="direction:rtl;line-height:200%;">
	<font face="vazir" size=4>
        صابر اول از همه می‌خواد یک انتشاراتی رو انتخاب کنه تا بره کتاباشو از اون‌جا بخره برای این کار احتیاج داره میانگین امتیاز هرکدوم از انتشاراتی‌ها رو بدونه تا بتونه بهترین ها رو فقط انتخاب کنه.
        <br>
        برای کمک به صابر، میانگین امتیاز کتاب‌های هر انتشاراتی را حساب کرده و در متغیر زیر بر حسب نام انتشاراتی ذخیره کنید.
    </font>
</div>



In [22]:
publisher_rating = df.groupby("publisher").average_rating.mean()
publisher_rating


publisher
18-Oct                           3.650000
1st Book Library                 3.580000
1st World Library                4.180000
A & C Black (Childrens books)    4.400000
A Harvest Book/Harcourt  Inc.    4.070000
                                   ...   
聯經出版事業股份有限公司                     4.400000
英文漢聲出版股份有限公司                     4.220000
角川書店 (Kadokawa Shoten)           3.815000
講談社                              4.283333
集英社                              4.237500
Name: average_rating, Length: 2290, dtype: float64


<br />
<div dir=rtl id="start" style="direction:rtl;line-height:200%;">
	<font face="vazir" size=4>
        صابر بعد از این که نگاهی به لیست انتشاراتی‌هایی که برایش انتخاب کرده بودید انداخت به این نتیجه رسید حوصله خوندن این همه کتاب رو نداره و بهتره فقط چند تا اطلاعات کلی در مورد هر نویسنده و انتشاراتی داشته باشه و اونا رو طوری جلوی دوستاش بگه که انگار همشونو خونده!
        <br>
        برای اولین قدم اون می‌خواد بدونه هر نویسنده چند کتاب با تعداد صفحات مشخصی نوشته. چیزی شبیه به جدول زیر:
        <br>
    </font>
</div>



<br />
<div dir=rtl id="start" style="direction:rtl;line-height:200%;">
	<font face="vazir" size=4>
      برای این کار ابتدا ستونی با نام num_pages_grp به دیتا فریم اضافه کنید که در آن تعداد صفحات کتاب را به صورت گسسته و در بازه‌های زیر نشان دهد.
    </font>
</div>
<250<br>
250-499<br>
500-749<br>
750-999<br>
>=1000<br>

<div dir=rtl id="start" style="direction:rtl;line-height:200%;">
	<font face="vazir" size=4>
      راهنمایی: از تابع np.select استفاده کنید.
    </font>
</div>

In [17]:
df['num_pages_grp']= np.select([df.num_pages>=1000,df.num_pages>=750,df.num_pages>=500,df.num_pages>=250,df.num_pages<250],[">=1000","750-999","500-749","250-499","<250"])


<br />
<div dir=rtl id="start" style="direction:rtl;line-height:200%;">
	<font face="vazir" size=4>
      مانند جدولی که نشان داده‌شد، تعداد کتاب‌های هر نویسنده برحسب صفحات کتاب را مشخص کنید.
    </font>
</div>

In [18]:
author_page = df.groupby(["author","num_pages_grp"]).size()
author_page


author                 num_pages_grp
A.B. Yehoshua          250-499          1
                       500-749          1
A.D.P. Briggs          250-499          1
A.E. Cunningham        <250             1
A.J. Jacobs            250-499          1
                                       ..
bell hooks             <250             5
Åsne Seierstad         250-499          3
Émile Zola             250-499          6
                       500-749          3
Éric-Emmanuel Schmitt  <250             1
Length: 5547, dtype: int64

<br />
<div dir=rtl id="start" style="direction:rtl;line-height:200%;">
	<font face="vazir" size=4>
    تبریک!
        <br>
        شما با موفقیت به پایان فصل پانداس رسیدید و توانستید به صابر برای کتاب‌خوان شدن کمک کنید.
    </font>
</div>

<br />
<div dir=rtl id="start" style="direction:rtl;line-height:200%;">
	<font face="vazir" size=4>
        در نهایت برای اطمینان از درست بودن پاسخ‌های خود، کد زیر را اجرا کرده و فایل result.zip را برای داوری ارسال کنید.
    </font>
</div>

In [19]:
import zlib
import zipfile


def compress(file_names):
    print("File Paths:")
    print(file_names)
    compression = zipfile.ZIP_DEFLATED
    with zipfile.ZipFile("result.zip", mode="w") as zf:
        for file_name in file_names:
            zf.write('./' + file_name, file_name, compress_type=compression)


df.to_csv('books_result.csv',index = False)
publisher_rating.to_csv('publisher_rating.csv',index = True)
author_page.to_csv("author_page.csv", index = True)

file_names = ["books_result.csv", "excersize 3.ipynb","publisher_rating.csv","author_page.csv"]
compress(file_names)

File Paths:
['books_result.csv', 'excersize 3.ipynb', 'publisher_rating.csv', 'author_page.csv']



<br />
<div dir=rtl id="start" style="direction:rtl;line-height:200%;">
	<font face="vazir" size
        سلاموری ارسال کنید.
  </p>
</div>